source/common/grpc/codec.h

```diff
@@ -0,0 +1,74 @@
+#pragma once
+
+#include "envoy/buffer/buffer.h"
+
+namespace Grpc {
+// Last bit for an expanded message without compression.
+const uint8_t GRPC_FH_DEFAULT = 0b0u;
+// Last bit for a compressed message.
+const uint8_t GRPC_FH_COMPRESSED = 0b1u;
+
+enum class CompressionAlgorithm { None, Gzip };
+
+struct Frame {
+  uint8_t flags;
+  uint32_t length;
+  Buffer::Instance* data;
+};
+
+class Encoder {
+public:
+  Encoder();
+
+  // Creates a new GRPC data frame with the given flags and length.
+  // @param flags supplies the GRPC data frame flags.
+  // @param length supplies the GRPC data frame length.
+  // @param output the buffer to store the encoded data, it's size must be 5.
+  void NewFrame(uint8_t flags, uint64_t length, uint8_t* output);
```

Comment:
using std::array<uint8_t, 5>? (would be good to enforce size 5 requirement here)